src/storage/mvcc/reader.rs

```diff
@@ -581,22 +580,22 @@ impl<S: Snapshot> MvccReader<S> {
             None => cursor.seek_to_first(&mut self.statistics.lock),
         };
         if !ok {
-            return Ok((vec![], None));
+            return Ok(vec![]);
         }
-        let mut locks = vec![];
+        let mut locks = Vec::with_capacity(limit);
         while cursor.valid() {
             let key = Key::from_encoded(cursor.key().to_vec());
             let lock = Lock::parse(cursor.value())?;
             if filter(&lock) {
-                locks.push((key.clone(), lock));
-                if limit > 0 && locks.len() >= limit {
-                    return Ok((locks, Some(key)));
+                locks.push((key, lock));
+                if limit > 0 && locks.len() == limit {
```

Comment:
do we need to check limit <= RESOLVE_LOCK_BATCH_SIZE, if not, the following lock getting may be wrong.

Reply:
@siddontang It belongs to different layers and should not happen. Otherwise you will see codes like this in the scheduler:

(pseudo code)

```
locks = MvccReader::scan_locks(limit = SOME_LIMIT_MORE_THAN_BATCH_SIZE);
check if we got RESOLVE_LOCK_BATCH_SIZE locks:
  yes: there are more locks
  no: we got all locks
```

As you can see, the mistake belongs to the caller (as code above), instead of MvccReader.

Reply:
if limit != RESOLVE_LOCK_BATCH_SIZE, any problem https://github.com/pingcap/tikv/pull/3350/files#diff-60c646555454e19a50f8de2b38f53522R565 here?

Reply:
ping @breeswish

Reply:
@siddontang Of course there will be problems. Because the logic is to get this number of locks and if we finally get less locks we will treat it as "scan end". If the scan limit you supplied is less, you must get less locks and in any condition it will be treated as "scan end".

Reply:
@siddontang Currently we have tests to ensure its correctness. And I think MvccReader does not have the responsibility to check it.

Reply:
I'm afraid that the existing tests are not strong enough..